An issue where the xlsx converted to csv contains an unprintable character like \0xa8 in its "value".

nitrocode opened the pull request:
This causes an issue when calling json.dumps() while updating a smartsheet.
The fix in place will check to see if there is an unprintable char by calling json.dumps() and if it fails it will iterate through all the values and converting only the ones that cause trouble.

Commit:
Aan issue where the xlsx converted to csv contains an unprintable character like \0xa8 in its "value".
This causes an issue when calling json.dumps() while updating a smartsheet.
The fix in place will check to see if there is an unprintable char by calling json.dumps() and if it fails it will iterate through all the values and converting only the ones that cause trouble.

stmcallister commented:
@nitrocode i'm running through some tests and I can't get json.dumps() to error with unprintable characters, using the code on the master branch of Smartsheet-Data-Tracker. Can you commit an example CSV file that causes the error?

nitrocode commented:
@stmcallister 

Unfortunately, I do not have that same file anymore and the unprintable character has been removed.

Here is a simple example of the issue we saw once the csv was imported by main.py

```
>>> mystr = {'test' : 'this will fail\xe1'}
>>> json.dumps(mystr)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python2.7/json/__init__.py", line 243, in dumps
    return _default_encoder.encode(obj)
  File "/usr/lib/python2.7/json/encoder.py", line 207, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/usr/lib/python2.7/json/encoder.py", line 270, in iterencode
    return _iterencode(o, 0)
UnicodeDecodeError: 'utf8' codec can't decode byte 0xe1 in position 12: unexpected end of data
```

I found out recently that there is a json.dumps() flag called `ensure_ascii` which can be set to false. This is another way around this issue however it won't remove the unprintable character. For us, that unprintable character is not something you want in your datasheet so we preferred to remove it.

```
>>> json.dumps(mystr, ensure_ascii=False)
'{"test": "this will fail\\u0000xa8\xe1"}'
```

or change the encoding

```
>>> json.dumps(mystr, encoding='latin1')
'{"test": "this will fail\\u00e1"}'
```

No idea how that unprintable character got in there.

Source: http://stackoverflow.com/questions/19872773/unicodedecodeerror-while-using-json-dumps

nitrocode committed:
Update main.py

nitrocode committed:
Added changes to move to the new 2.0 API

All I had to do was change app.json to use the 2.0 api url and convert "row" and "sheet" to their plural forms within the URLs for retrieving a row and updating a row.

stmcallister commented:
While I appreciate the effort, if we can reproduce the error, we can't test to see if the fix works. I'm going to close this pull request until we can reproduce the error and then verify that the fix works.
